### PR TITLE
Audio

### LINKED ISSUE #181: Design audio interface
Create an interface between audio and engine.

### PR DESCRIPTION
**Pull request by NiceGuyAndersson**
First draft of how the Steam Audio implementation will look. I _think_ it's complete, but I expect to learn a lot more about SA when it's time to actually write the code.

Next up is to replace OpenAL with something else, next sprint!?

Fixes #181

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Audio/SteamAudio.hpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+class SteamAudio {
+public:
+	/// Processes the audio with both indirect and direct methods and mixes them together.
+	IPLAudioBuffer process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* PlayerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
+
+private:
+	SteamAudioDirectRenderer directRenderer;
+	SteamAudioIndirectRenderer indirectRenderer;
+	IPLhandle* environmentalRenderer;
+};
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Missing newline at end of file.

File: src/Engine/Audio/SteamAudio.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+#include "SteamAudioDirectRenderer.hpp"
+#include "SteamAudioIndirectRenderer.hpp"
+
+
+class SteamAudio {
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Document class.

File: src/Engine/Audio/SteamAudio.hpp [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+public:
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Indent modifiers.

File: src/Engine/Audio/SteamAudio.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+	/// Processes the audio with both indirect and direct methods and mixes them together.
+	IPLAudioBuffer process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* PlayerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
+
+private:
+	SteamAudioDirectRenderer directRenderer;
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Use spaces, not tabs.

File: src/Engine/Audio/SteamAudio.hpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+#include "SteamAudioDirectRenderer.hpp"
+#include "SteamAudioIndirectRenderer.hpp"
+
+
+class SteamAudio {
+public:
+	/// Processes the audio with both indirect and direct methods and mixes them together.
+	IPLAudioBuffer process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* PlayerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Document parameters and return values.
Functions should be named in PascalCase, so Process.

File: src/Engine/Audio/SteamAudioDirectRenderer.hpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include <phonon.h>
+
+class SteamAudioDirectRenderer {
+public:
+
+	SteamAudioDirectRenderer();
+
+	SteamAudioDirectRenderer(IPLContext context, IPLhandle* environment);
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioDirectRenderer.hpp]: Document parameters.

File: src/Engine/Audio/SteamAudioDirectRenderer.hpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include <phonon.h>
+
+class SteamAudioDirectRenderer {
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioDirectRenderer.hpp]: Document class.

File: src/Engine/Audio/SteamAudioIndirectRenderer.hpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+#include <phonon.h>
+
+class SteamAudioIndirectRenderer {
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioIndirectRenderer.hpp]: Document class

File: src/Engine/Audio/SteamAudioInterface.hpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
+#pragma once
+#include <phonon.h>
+#include <vector>
+#include "SteamAudio.hpp"
+
+class SteamAudioInterface {
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioInterface.hpp]: Document class.

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [SoundManager.cpp#L33](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c8f86e51a0c769001d9c24510c23898e1a48d26f/src/Engine/Manager/SoundManager.cpp#L33): The function 'SetVolume' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [SoundManager.cpp#L38](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c8f86e51a0c769001d9c24510c23898e1a48d26f/src/Engine/Manager/SoundManager.cpp#L38): The function 'GetVolume' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**